cpp/teensy/lib/multilateration/multilateration.cpp

```diff
@@ -0,0 +1,125 @@
+        return Status;
+    }
+
+    arm_matrix_instance_f32 AtransXAinvXAtrans;
+    AtransXAinvXAtrans.numCols = NUM_HYDROPHONES - 1;
+    AtransXAinvXAtrans.numRows = NUM_HYDROPHONES - 1;
+    AtransXAinvXAtrans.pData =
+        new float32_t[(NUM_HYDROPHONES - 1) * (NUM_HYDROPHONES - 1)];
+    const arm_matrix_instance_f32* pAtransAinv = &AtransXAinv;
+    Status = arm_mat_mult_f32(pAtransAinv, pAtrans, &AtransXAinvXAtrans);
+    if (Status != ARM_MATH_SUCCESS) {
+        return Status;
+    }
+
+    const arm_matrix_instance_f32* pAtransXAinvXAtrans = &AtransXAinvXAtrans;
+    Status = arm_mat_mult_f32(pAtransXAinvXAtrans, pB, pResult);
+    if (Status != ARM_MATH_SUCCESS) {
+        return Status;
+    }
+
+    delete[] Atrans.pData;
+    delete[] AtransXA.pData;
+    delete[] AtransXAinv.pData;
+    delete[] AtransXAinvXAtrans.pData;
+
+    return Status;
+}
+
+void initialComputationA(float32_t* AData,
+                         HydrophonePositions HydrophonePositions[]) {
+    for (int i = 0; i < (NUM_HYDROPHONES - 1); i++) {
+        *(AData + i * (NUM_HYDROPHONES - 1) + 0) =
+            HydrophonePositions[0].PosX - HydrophonePositions[i + 1].PosX;
+        *(AData + i * (NUM_HYDROPHONES - 1) + 1) =
+            HydrophonePositions[0].PosY - HydrophonePositions[i + 1].PosY;
+        *(AData + i * (NUM_HYDROPHONES - 1) + 2) =
+            HydrophonePositions[0].PosZ - HydrophonePositions[i + 1].PosZ;
+        *(AData + i * (NUM_HYDROPHONES - 1) + 3) = 0.0;
+    }
+}
+
+void compute_A(int32_t TdoaArray[], float32_t* AData) {
+    for (int i = 0; i < (NUM_HYDROPHONES - 1); i++) {
+        *(AData + i * (NUM_HYDROPHONES - 1) + 3) =
+            TdoaArray[i] * SOUND_SPEED / SAMPLING_FREQ;
+    }
+}
+
+void compute_B(int32_t TdoaArray[], HydrophonePositions HydrophonePositions[],
+               float32_t* Bdata) {
+    for (int i = 0; i < (NUM_HYDROPHONES - 1); i++) {
+        *(Bdata + i) =
+            0.5 * (pow(HydrophonePositions[0].PosX, 2) -
+                   pow(HydrophonePositions[i + 1].PosX, 2) +
+                   pow(HydrophonePositions[0].PosY, 2) -
+                   pow(HydrophonePositions[i + 1].PosY, 2) +
+                   pow(HydrophonePositions[0].PosZ, 2) -
+                   pow(HydrophonePositions[i + 1].PosZ, 2) +
+                   pow(TdoaArray[i] * SOUND_SPEED / SAMPLING_FREQ, 2));
+    }
+}
+
+void initHydrophonePositions(HydrophonePositions* hydrophonePositions) {
+    HydrophonePositions Hyd0 = {HYD_0_X_POS, HYD_0_Y_POS, HYD_0_Z_POS};
+    HydrophonePositions Hyd1 = {HYD_1_X_POS, HYD_1_Y_POS, HYD_1_Z_POS};
+    HydrophonePositions Hyd2 = {HYD_2_X_POS, HYD_2_Y_POS, HYD_2_Z_POS};
+    HydrophonePositions Hyd3 = {HYD_3_X_POS, HYD_3_Y_POS, HYD_3_Z_POS};
+    HydrophonePositions Hyd4 = {HYD_4_X_POS, HYD_4_Y_POS, HYD_4_Z_POS};
+
+    hydrophonePositions[0] = Hyd0;
+    hydrophonePositions[1] = Hyd1;
+    hydrophonePositions[2] = Hyd2;
+    hydrophonePositions[3] = Hyd3;
+    hydrophonePositions[4] = Hyd4;
+}
```

Comment:
Should probably be a constant array of `const Position HydrophonePositions[5]`.

Reply:
What would be a good way to do this? I could allocate an array inside the init function, and use a return value, but this is not very pretty.

Reply:
This is a constant at this point.
```C++
const Position hydrophonePositions[5] = {
    {..., ..., ...},
    .    ...
    {..., ..., ...},
};
```
Is there a reason this doesn't work?